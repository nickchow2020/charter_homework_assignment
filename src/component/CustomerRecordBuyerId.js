import { useContext } from 'react';
import CustomerRecord from './CustomerRecord';
import helper from '../helper';
import generalContext from '../generalContext';
const CustomerRecordBuyerId = ({ month, buyerId }) => {
  const { transactionData } = useContext(generalContext);
  const transaction = transactionData[month][buyerId];
  const totalPoint = helper.calculateTotalRewardPointPerMonth(transaction);

  return (
    <table style={{ width: '100%', textAlign: 'center', marginBottom: '2rem' }}>
      <thead>
        <tr>
          <th>BuyerId</th>
          <th>TransitionNumber</th>
          <th>Name</th>
          <th>PurchaseItem</th>
          <th>Price</th>
          <th>RewardPoints</th>
          <th>Date</th>
        </tr>
      </thead>
      <tbody>
        {transaction.map((record) => (
          <CustomerRecord record={record} key={record.transactionNumber} />
        ))}

        <tr>
          <th colSpan={5} style={{ textAlign: 'right' }}>
            Total Points:
          </th>
          <td>{`${totalPoint}`}</td>
          <td></td>
        </tr>
      </tbody>
    </table>
  );
};

/**
 *
 * Data Structure for transactionData
 * @example
 * {
 *  May:{
 *    buyerId1:[{...buyer1},{...buyer1},{...buyer1},{...buyer1}]
 *    buyerId2:[{...buyer2},{...buyer2},{...buyer2},{...buyer2}]
 *    buyerId3:[{...buyer3},{...buyer3},{...buyer3},{...buyer3}]
 *    }
 *  Jun:{
 *    buyerId1:[{...buyer1},{...buyer1},{...buyer1},{...buyer1}]
 *    buyerId2:[{...buyer2},{...buyer2},{...buyer2},{...buyer2}]
 *    buyerId3:[{...buyer3},{...buyer3},{...buyer3},{...buyer3}]
 *    }
 *  Jul:{
 *    buyerId1:[{...buyer1},{...buyer1},{...buyer1},{...buyer1}]
 *    buyerId2:[{...buyer2},{...buyer2},{...buyer2},{...buyer2}]
 *    buyerId3:[{...buyer3},{...buyer3},{...buyer3},{...buyer3}]
 *    }
 * }
 */

export default CustomerRecordBuyerId;
