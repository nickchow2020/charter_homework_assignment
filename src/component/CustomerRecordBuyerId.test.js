import { render, screen } from '@testing-library/react';
import CustomerRecordBuyerId from './CustomerRecordBuyerId';
import GeneralContext from '../generalContext';

describe('<CustomerRecordBuyerId />', () => {
  const mockData = {
    May: {
      test1User: [
        {
          buyerId: 'test1User',
          itemId: '4f8c-9a93-b404057ca6ad',
          itemName: 'v-T5rIt^&t',
          rewardPoints: 30,
          transactionDate: 'May 05 2022',
          transactionNumber: 'c55-ae17-a9f22f2100a3',
        },
        {
          buyerId: 'test1User',
          itemId: 'b498a452-75e9-444f',
          itemName: '&A{H2@$v5]',
          rewardPoints: 110,
          transactionDate: 'May 15 2022',
          transactionNumber: 'e0903f84',
        },
      ],
      test2User: [
        {
          buyerId: 'test2User',
          itemId: 'be011917-1c7a-4f8c-9a93-b',
          itemName: 'v-T5rIt^&t',
          rewardPoints: 30,
          transactionDate: 'May 05 2022',
          transactionNumber: 'bd6b53eb-8d5a-4c5',
        },
        {
          buyerId: 'test2User',
          itemId: '444f-be07-af6322bc34b9',
          itemName: '&A{H2@$v5]',
          rewardPoints: 110,
          transactionDate: 'May 15 2022',
          transactionNumber: 'e0903f84-2611-45e0-',
        },
      ],
    },
  };
  it('should match snapshot', () => {
    const { asFragment } = render(
      <GeneralContext.Provider value={{ transactionData: mockData }}>
        <CustomerRecordBuyerId month='May' buyerId='test1User' />
      </GeneralContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display total points correctly', () => {
    render(
      <GeneralContext.Provider value={{ transactionData: mockData }}>
        <CustomerRecordBuyerId month='May' buyerId='test1User' />
      </GeneralContext.Provider>
    );

    const displayTotal = screen.getByText('140');

    // 30 + 110
    expect(displayTotal).toBeInTheDocument();
  });
});
