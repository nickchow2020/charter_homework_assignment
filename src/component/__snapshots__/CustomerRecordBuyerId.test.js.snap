// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomerRecordBuyerId /> should match snapshot 1`] = `
<DocumentFragment>
  <table
    style="width: 100%; text-align: center; margin-bottom: 2rem;"
  >
    <thead>
      <tr>
        <th>
          BuyerId
        </th>
        <th>
          TransitionNumber
        </th>
        <th>
          Name
        </th>
        <th>
          PurchaseItem
        </th>
        <th>
          Price
        </th>
        <th>
          RewardPoints
        </th>
        <th>
          Date
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          test1User
        </td>
        <td>
          c55-ae17-a9f22f2100a3
        </td>
        <td />
        <td />
        <td />
        <td>
          30
        </td>
        <td>
          May 05 2022
        </td>
      </tr>
      <tr>
        <td>
          test1User
        </td>
        <td>
          e0903f84
        </td>
        <td />
        <td />
        <td />
        <td>
          110
        </td>
        <td>
          May 15 2022
        </td>
      </tr>
      <tr>
        <th
          colspan="5"
          style="text-align: right;"
        >
          Total Points:
        </th>
        <td>
          140
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
